File: frost-taproot/src/poly.rs
// Mirrors ref/frost/src/lib/poly.ts
// Polynomial evaluation and Lagrange interpolation over the secp256k1 scalar field.

use k256::Scalar;

use crate::ecc::util::{scalar_from_bytes, scalar_invert};
use crate::util::assert;
use crate::Error;

/// Evaluate a polynomial at x using Horner's method.
/// Coefficients are in ascending order: coeffs[0] + coeffs[1]*x + coeffs[2]*x^2 + ...
/// Mirrors `evaluate_x` in the TS implementation.
pub fn evaluate_x(coeffs: &[Scalar], x: Scalar) -> Result<Scalar, Error> {
    if x == Scalar::ZERO {
        return Err(Error::Assertion("x is zero".to_string()));
    }

    let mut value = Scalar::ZERO;
    for coeff in coeffs.iter().rev() {
        value = value * x + coeff;
    }
    Ok(value)
}

/// Interpolate a polynomial at x=0 (the root) using Lagrange interpolation.
/// Points are (x, y) pairs of Scalars.
/// Mirrors `interpolate_root` in the TS implementation.
pub fn interpolate_root(points: &[(Scalar, Scalar)]) -> Result<Scalar, Error> {
    let xs: Vec<Scalar> = points.iter().map(|(x, _)| *x).collect();
    let mut p = Scalar::ZERO;
    for (x, y) in points {
        let delta = interpolate_x(&xs, *x)?;
        p = p + delta * y;
    }
    Ok(p)
}

/// Compute the Lagrange basis polynomial value at x=0 for the given x-coordinate,
/// relative to the set of x-coordinates L.
/// Mirrors `interpolate_x` in the TS implementation.
pub fn interpolate_x(l: &[Scalar], x: Scalar) -> Result<Scalar, Error> {
    assert::is_included(l, &x)?;
    assert::is_unique_set(l)?;

    let mut numerator = Scalar::ONE;
    let mut denominator = Scalar::ONE;

    for &x_j in l {
        if x_j == x {
            continue;
        }
        numerator = numerator * x_j;
        denominator = denominator * (x_j + (-x)); // x_j - x
    }

    let inv = scalar_invert(&denominator)?;
    Ok(numerator * inv)
}

/// Compute the Lagrange coefficient for participant P relative to members L,
/// evaluated at point x.
/// Mirrors `calc_lagrange_coeff` in the TS implementation.
pub fn calc_lagrange_coeff(l: &[Scalar], p: Scalar, x: Scalar) -> Result<Scalar, Error> {
    assert::is_unique_set(l)?;

    let mut numerator = Scalar::ONE;
    let mut denominator = Scalar::ONE;

    for &x_j in l {
        if x_j == p {
            continue;
        }
        numerator = numerator * (x + (-x_j)); // x - x_j
        denominator = denominator * (p + (-x_j)); // p - x_j
    }

    let inv = scalar_invert(&denominator)?;
    Ok(numerator * inv)
}

/// Convert a u32 participant index to a Scalar.
pub fn index_to_scalar(idx: u32) -> Scalar {
    scalar_from_bytes(&{
        let mut b = [0u8; 32];
        b[28..].copy_from_slice(&idx.to_be_bytes());
        b
    })
}

File: frost-taproot-go/poly/poly.go
// Package poly provides polynomial evaluation and Lagrange interpolation.
package poly

import (
	"fmt"
	"math/big"

	"github.com/frost-taproot/frost-taproot-go/ecc"
)

// EvaluateX evaluates a polynomial at x using Horner's method.
func EvaluateX(coeffs []*big.Int, x *big.Int) (*big.Int, error) {
	if x.Sign() == 0 {
		return nil, fmt.Errorf("x is zero")
	}

	value := big.NewInt(0)
	for i := len(coeffs) - 1; i >= 0; i-- {
		value = ecc.ModN(new(big.Int).Add(ecc.ScalarMul(value, x), coeffs[i]))
	}
	return value, nil
}

// InterpolateRoot interpolates at x=0 using Lagrange interpolation.
func InterpolateRoot(points [][2]*big.Int) (*big.Int, error) {
	xs := make([]*big.Int, len(points))
	for i, p := range points {
		xs[i] = p[0]
	}

	p := big.NewInt(0)
	for _, point := range points {
		x := point[0]
		y := point[1]
		delta, err := InterpolateX(xs, x)
		if err != nil {
			return nil, err
		}
		p = ecc.ScalarAdd(p, ecc.ScalarMul(delta, y))
	}
	return p, nil
}

// InterpolateX computes Lagrange basis at x=0.
func InterpolateX(l []*big.Int, x *big.Int) (*big.Int, error) {
	if !isIncluded(l, x) {
		return nil, fmt.Errorf("x not included in set")
	}
	if err := IsUniqueSet(l); err != nil {
		return nil, err
	}

	numerator := big.NewInt(1)
	denominator := big.NewInt(1)

	for _, xj := range l {
		if xj.Cmp(x) == 0 {
			continue
		}
		numerator = ecc.ScalarMul(numerator, xj)
		denominator = ecc.ScalarMul(denominator, ecc.ScalarAdd(xj, ecc.ScalarNeg(x)))
	}

	inv, err := ecc.ScalarInvert(denominator)
	if err != nil {
		return nil, err
	}
	return ecc.ScalarMul(numerator, inv), nil
}

// CalcLagrangeCoeff computes Lagrange coefficient.
func CalcLagrangeCoeff(l []*big.Int, p, x *big.Int) (*big.Int, error) {
	if err := IsUniqueSet(l); err != nil {
		return nil, err
	}

	numerator := big.NewInt(1)
	denominator := big.NewInt(1)

	for _, xj := range l {
		if xj.Cmp(p) == 0 {
			continue
		}
		numerator = ecc.ScalarMul(numerator, ecc.ScalarAdd(x, ecc.ScalarNeg(xj)))
		denominator = ecc.ScalarMul(denominator, ecc.ScalarAdd(p, ecc.ScalarNeg(xj)))
	}

	inv, err := ecc.ScalarInvert(denominator)
	if err != nil {
		return nil, err
	}
	return ecc.ScalarMul(numerator, inv), nil
}

// IndexToScalar converts uint32 to scalar.
func IndexToScalar(idx uint32) *big.Int {
	var b [32]byte
	b[28] = byte(idx >> 24)
	b[29] = byte(idx >> 16)
	b[30] = byte(idx >> 8)
	b[31] = byte(idx)
	return ecc.ScalarFromBytes(b)
}

// IsUniqueSet checks if all elements are unique.
func IsUniqueSet(l []*big.Int) error {
	seen := make(map[string]bool)
	for _, x := range l {
		key := x.Text(16)
		if seen[key] {
			return fmt.Errorf("duplicate element in set")
		}
		seen[key] = true
	}
	return nil
}

func isIncluded(l []*big.Int, x *big.Int) bool {
	for _, v := range l {
		if v.Cmp(x) == 0 {
			return true
		}
	}
	return false
}
